typegen/strats/stub.py

```diff
@@ -1,5 +1,92 @@
-from typegen.strats.gen import TypeHintGenerator
+import pathlib
+import libcst as cst
+import tempfile
+from mypy import stubgen
+from typegen.strats.inline import InlineGenerator
 
 
-class StubFileGenerator(TypeHintGenerator):
-    ident = "stub"
\ No newline at end of file
+class _ImportUnionTransformer(cst.CSTTransformer):
+    def __init__(self):
+        self.requires_union_import = False
+
+    def leave_Module(self, original_node: cst.Module, updated_node: cst.Module):
+        if self.requires_union_import:
+            typing_union_import = cst.ImportFrom(
+                module=cst.Name(value="typing"),
+                names=[cst.ImportAlias(cst.Name("Union"))],
+            )
+
+            typings_line = cst.SimpleStatementLine([typing_union_import])
+            new_body = [typings_line] + list(updated_node.body)  # type: ignore
+            # Note: Same code as in typegen/strats/imports.py
+
+            return updated_node.with_changes(body=new_body)
+        return updated_node
+
+    def visit_Param(self, node: cst.Param) -> bool | None:
+        if not hasattr(node, "annotation") or node.annotation is None:
+            return True
+        self._check_annotation_union(node.annotation.annotation)
+        return True
+
+    def visit_FunctionDef_returns(self, node: cst.FunctionDef) -> None:
+        if node.returns:
+            self._check_annotation_union(node.returns.annotation)
+
+    def visit_AnnAssign(self, node: cst.AnnAssign) -> bool | None:
+        self._check_annotation_union(node.annotation.annotation)
+        return True
+
+    def _check_annotation_union(self, node: cst.CSTNode | None) -> None:
+        if self.requires_union_import:
+            return
+        if isinstance(node, cst.Subscript):
+            if isinstance(node.value, cst.Name) and node.value.value == "Union":
+                self.requires_union_import = True
+
+
+class StubFileGenerator(InlineGenerator):
```

Comment:
The `StubFileGenerator` is not an inline generator; the types are stored next to the original source code.
Inherit from `TypeHintGenerator` instead, and replace the calls to `super()._store_hinted_ast` appropriately

Reply:
Refactored and co-authored together